kb/2025-10-07/velero-backup-filesystem-freeze.md

```diff
@@ -0,0 +1,193 @@
+description: 'Learn how to implement filesystem freeze for External CSI Storage Backup with Velero to ensure data consistency.'
+authors:
+  - name: Webber Huang
+    title: Senior Software Engineer
+    url: https://github.com/WebberHuang1118
+    image_url: https://github.com/WebberHuang1118.png
+tags: [backup, csi, storage, velero, filesystem-freeze]
+hide_table_of_contents: false
+---
+
+In this [Harvester Knowledge Base article](https://harvesterhci.io/kb/2025/05/26/velero-backup-restore), [Ivan Sim](https://github.com/ihcsim) provided comprehensive guidance on using [Velero](https://velero.io/) to perform backup and restore operations for VMs with external storage in Harvester.
+
+However, in certain scenarios, users may require the VM filesystem to be quiesced during Velero backup creation to prevent data corruption, especially when the VM is experiencing heavy I/O operations.
+
+This article describes how to customize [Velero Backup Hooks](https://velero.io/docs/v1.17/backup-hooks/) to implement filesystem freeze during Velero backup processing, ensuring data consistency in the backup content.
+
+## Background Knowledge
+
+KubeVirt's [virt-freezer](https://github.com/kubevirt/kubevirt/blob/main/docs/freeze.md#virt-freezer) provides a mechanism to freeze and thaw guest filesystems. This capability can be leveraged to ensure filesystem consistency during VM backups. However, certain prerequisites must be met for filesystem freeze/thaw operations to function properly:
+
+### Prerequisites for Filesystem Freeze
+
+* **QEMU Guest Agent must be enabled** in the guest VM
+  * Verify this by checking if the VMI has **AgentConnected** in its status
+* **Guest VM must be properly configured** for related libvirt commands
+  * When **virt-freezer** is triggered, KubeVirt communicates with the **QEMU Guest Agent** via libvirt commands such as **guest-fsfreeze-freeze**
+  * The guest agent translates these commands to OS-specific calls:
+    * **Linux systems**: Uses **fsfreeze** syscalls
+    * **Windows systems**: Uses **VSS** (Volume Shadow Copy Service) APIs
+
+### Common Configuration Challenges
+
+Based on Harvester project experience, some guest operating systems require additional configuration:
+
+* **Linux distributions** (e.g., RHEL, SLE Micro): May lack sufficient permissions for filesystem freeze operations by default, requiring custom policies
+* **Windows guests**: Require the **VSS** service to be enabled for filesystem freeze functionality
+
+> **Important**: Filesystem freeze/thaw functionality depends on guest VM configuration, which is outside Harvester's control. Users are responsible for ensuring compatibility before implementing Velero backup hooks with filesystem freeze.
+
+### Verifying Filesystem Freeze Compatibility
+
+To confirm that your VM supports filesystem freeze operations:
+
+1. Access the virtual machine's virt-launcher `compute` container:
+
+   ```bash
+   POD=$(kubectl get pods -n default \
```

Comment:
Suggest using `<VM namespace>` and `<VM name>` in the examples here to be consistent with the following example command.